promql/engine_test.go

```diff
@@ -542,234 +711,496 @@ load 10s
-				},
-				nil,
-			},
-			Start: time.Unix(10, 0),
+			Start:      time.Unix(10, 0),
 		},
 		{
-			Query:      "metric[20s]",
-			MaxSamples: 0,
-			Result: Result{
-				ErrTooManySamples(env),
-				nil,
-				nil,
-			},
-			Start: time.Unix(10, 0),
+			Query:      "metric[20s] @ 10",
+			MaxSamples: 2,
+			Start:      time.Unix(0, 0),
 		},
 		// Range queries.
 		{
 			Query:      "1",
 			MaxSamples: 3,
-			Result: Result{
-				nil,
-				Matrix{Series{
-					Points: []Point{{V: 1, T: 0}, {V: 1, T: 1000}, {V: 1, T: 2000}},
-					Metric: labels.FromStrings()},
-				},
-				nil,
-			},
-			Start:    time.Unix(0, 0),
-			End:      time.Unix(2, 0),
-			Interval: time.Second,
+			Start:      time.Unix(0, 0),
+			End:        time.Unix(2, 0),
+			Interval:   time.Second,
 		},
 		{
 			Query:      "1",
-			MaxSamples: 0,
-			Result: Result{
-				ErrTooManySamples(env),
-				nil,
-				nil,
-			},
-			Start:    time.Unix(0, 0),
-			End:      time.Unix(2, 0),
-			Interval: time.Second,
+			MaxSamples: 3,
+			Start:      time.Unix(0, 0),
+			End:        time.Unix(2, 0),
+			Interval:   time.Second,
 		},
 		{
 			Query:      "metric",
 			MaxSamples: 3,
-			Result: Result{
-				nil,
-				Matrix{Series{
-					Points: []Point{{V: 1, T: 0}, {V: 1, T: 1000}, {V: 1, T: 2000}},
-					Metric: labels.FromStrings("__name__", "metric")},
-				},
-				nil,
-			},
-			Start:    time.Unix(0, 0),
-			End:      time.Unix(2, 0),
-			Interval: time.Second,
+			Start:      time.Unix(0, 0),
+			End:        time.Unix(2, 0),
+			Interval:   time.Second,
 		},
 		{
 			Query:      "metric",
-			MaxSamples: 2,
-			Result: Result{
-				ErrTooManySamples(env),
-				nil,
-				nil,
-			},
-			Start:    time.Unix(0, 0),
-			End:      time.Unix(2, 0),
-			Interval: time.Second,
+			MaxSamples: 3,
+			Start:      time.Unix(0, 0),
+			End:        time.Unix(10, 0),
+			Interval:   5 * time.Second,
 		},
 		{
-			Query:      "metric",
+			Query:      "rate(bigmetric[1s])",
+			MaxSamples: 1,
+			Start:      time.Unix(0, 0),
+			End:        time.Unix(10, 0),
+			Interval:   5 * time.Second,
+		},
+		{
+			// Result is duplicated, so @ also produces 3 samples.
+			Query:      "metric @ 10",
 			MaxSamples: 3,
-			Result: Result{
-				nil,
-				Matrix{Series{
-					Points: []Point{{V: 1, T: 0}, {V: 1, T: 5000}, {V: 2, T: 10000}},
-					Metric: labels.FromStrings("__name__", "metric")},
-				},
-				nil,
-			},
-			Start:    time.Unix(0, 0),
-			End:      time.Unix(10, 0),
-			Interval: 5 * time.Second,
+			Start:      time.Unix(0, 0),
+			End:        time.Unix(10, 0),
+			Interval:   5 * time.Second,
 		},
 		{
-			Query:      "metric",
-			MaxSamples: 2,
-			Result: Result{
-				ErrTooManySamples(env),
-				nil,
-				nil,
-			},
-			Start:    time.Unix(0, 0),
-			End:      time.Unix(10, 0),
-			Interval: 5 * time.Second,
+			// With @ modifier, the result is duplicated for the time range, hence it counts.
+			// The selector itself counts does not count for any sample.
```

Comment:
This doesn't parse

Reply:
Will update the comment. I meant the iterator for series lazily loads the samples and does not take space upfront.

Reply:
Unlike subqueries